### ISSUE TITLE
parallelize calculation of duplicate extends

### ISSUE THREAD

**Issue by lpirl** [markfasheh/duperemove#69]
I certainly do not reveal a dark secret when I report that "calculating duplicate extents" takes veeery long. Even "this may take some time" is a bit understated.

This issue feature-requests if there any room for parallelization.

**Comment by lpirl** [comment]
I glanced at `dbfile_load_hashes_bloom` in `dbfile.c` and it seems that there is one query per almost-dupes-hash.
Could we ask SQLite to give us only digests that occur more than once and iterate over those?
Maybe with a statement like:

<pre>select digest from hashes group by digest having count(*) > 1;</pre>

(takes 12.2 seconds for 10.6M hashes on a i5-4430 @ 3GHz)

**Comment by markfasheh** [comment]
So we have an item on our list of development tasks - search for "Multi-threaded extent search":

https://github.com/markfasheh/duperemove/wiki/Development-Tasks

By extent search, I mean 'find_all_dupes()' - it is entirely CPU bound. There is no I/O, nor is there memory allocation.

The last time I took a crack at this, most of what I tried wound up slowing things down. From memory:

   - first pass at threading it added maybe one mutex which actually blew up the time to search significantly.
   - I tried replacing some of the locking with atomic updates
       - obviously, they are extremely tricky to use
       - AFAIR gcc uses a full memory barrier for these
          - that turned out to be slow too

I'm not giving up on it yet or anything but it's certainly not an easy change. If you have any ideas or things to try by all means please do. I'll happily help out.

**Comment by markfasheh** [comment]
Regarding dbfile_load_hashes_bloom(), the first pass reproduces the logic we had when using flat files. There wasn't much of a change in performance from the flat file. Now that we have a DB backend though, optimizations like yours are exactly what I'd like to see.

Just keep in mind this is all going to take time. If you have the time / skills to make the change to our dbfile code feel free, I appreciate contributions and try to put real effort into reviewing and including them.


Lastly, can we quantify what you mean by 'very long'? A couple questions - what command line are you running, what version are you running with and can you describe your data set by estimating num files, average size of files, and total size of the file set?

**Comment by markfasheh** [comment]
Btw, closing for now to keep the issue list clean.

**Comment by lpirl** [comment]
Thanks for your interesting explanations.

By 'very long', I mean 136 hours CPU time for the run described in #56. I think I'll kill that soon…

Of course I see that this all needs the resource that's restricted over here too: time. I'd have to get used to low-level programming again and to figure out how your data structures/logic work/s so it is probably less efficient if I do the task. However, I'll try if I find the time ;)

Do you think you could include the idea about `dbfile_load_hashes_bloom` in the [development tasks](https://github.com/markfasheh/duperemove/wiki/Development-Tasks)? Suggestion:

- [ ] Use capabilities of SQL for efficiency
  - `dbfile_load_hashes_bloom`: query for digests that occur more than once and iterate over those (e.g. `select digest from hashes group by digest having count(*) > 1;`)

**Comment by markfasheh** [comment]
Good idea, I've got it in there now, thanks.

**Comment by lpirl** [comment]
Nice. Thanks.

**Comment by lpirl** [comment]
@markfasheh 
Because of a discussion in another thread, I read `dbfile_load_hashes_bloom` again and somehow think that the above proposed fix wouldn't be too hard.
It would be nice if you could verify my understanding of what that function does.

It iterates over the `d_tree` to looks all hashes therein up in the database to finally call `insert_hashed_block` with the hash, the file and so on.

Now, if we would get all duplicate hashes from the database directly, we could simply loop over those, and for each row look up the file and call `insert_hashed_block`.

In other words, `insert_hashed_block` is independent of the loop over the `d_tree`.

Is that correct?

**Comment by juliantaylor** [comment]
before spending time on trying to optimize that function further, make sure to profile it again after the index fix I mentioned there.
With the fix you can expect it to be 10000-100000 times faster than before depending on how many hashes you have. This speedup could be already good enough.
This will likely make the runtime of the function marginal compared to `walk_dupe_hashes` which is O(N^2) in number of duplicate hashes.

Do you have compressed files in your dataset? Larger compressed files in btrfs appear to be built from from thousands of smaller extents, a duplicate here will dominate the runtime in `walk_dupe_hashes`

**Comment by lpirl** [comment]
@juliantaylor certainly I will - thanks for the fix. And yes, I have compressed files only.

**Comment by lpirl** [comment]
Additionally: can we even get rid of the bloom when using SQL to get duplicates?

**Comment by markfasheh** [comment]
Most likely, yes though I agree with juliantaylor that we should profile with the fix first.
